Component docs update (#197)
enhance-dev/enhance.dev: app/docs/md/conventions/components.md

title: Components
---

> When building applications with Enhance we recommend starting with [Elements](/docs/elements). Elements are the perfect solution when you need server side rendering and don't require client side interactivity. Let's face it the majority of your application is presentation only so why would you want to author that with client side JavaScript.

Reviewer: ```suggestion
When building UI elements for Enhance applications, we recommend starting with [Elements](/docs/elements). Elements are the perfect solution when you need server side rendering and don't require client side interactivity. We find that most applications are composed with a majority of strictly presentational components, and authoring that code with client side JavaScript can be counterproductive.
```

> When you need to provide client side interactivity Components are what to reach for. They wrap your HTML, CSS and JavaScript in a portable web component. They are portable since these components can render themselves client side outside of Enhance applications. However, when they are included in an Enhance application they gain the super power of being rendered on the server and then hydrate themselves on the client. Components live in the `app/components/` folder in Enhance projects.

Reviewer: ```suggestion
When you need to provide client side interactivity, Components are what to reach for. They wrap your HTML, CSS and JavaScript in a portable web component. They are ‘portable’ because these components can render themselves client side outside of Enhance applications. However, when they are used in an Enhance application, they gain the super power of being rendered on the server and then hydrating themselves on the client. Components live in the `app/components/` folder in Enhance projects.
```

> When building UI elements for Enhance applications, we recommend starting with [Elements](/docs/elements). Elements are the perfect solution when you need server side rendering and don't require client side interactivity. We find that most applications are composed with a majority of strictly presentational components, and authoring that code with client side JavaScript can be counterproductive.
> 
> When you need to provide client side interactivity, Components are what to reach for. They wrap your HTML, CSS and JavaScript in a portable web component. They are ‘portable’ because these components can render themselves client side outside of Enhance applications. However, when they are used in an Enhance application, they gain the super power of being rendered on the server and then hydrating themselves on the client. Components live in the `app/components/` folder in Enhance projects.

Reviewer: On reading this through a second time, I worry that this might make it seem like Elements are incapable of client side capabilities. I know we want to provide concrete guidance for people, I just also don't want to minimize existing functionality that also works really well.

What are your thoughts on something like this?

```suggestion
When building UI elements for Enhance applications, we recommend starting with [Elements](/docs/elements). Elements are the perfect solution for starting with server side rendering and, optionally, iteratively upgrading to client side interaction. We find that most applications are composed with a majority of strictly presentational components, and authoring that code with client side JavaScript can be counterproductive.

Components, meanwhile, provide several shortcuts to client side interactivity while maintaining server side rendering. They wrap your HTML, CSS and JavaScript in a portable web component. They are ‘portable’ because these components can render themselves client side outside of Enhance applications. However, when they are used in an Enhance application, they gain the super power of being rendered on the server and then hydrating themselves on the client. Components live in the `app/components/` folder in Enhance projects.
```

## Lifecycle

Let's look at the lifecycle of a component we'll call `DeleteButton`. This component will act like a normal submit button when JavaScript is not available but will override the default behavior of the submit button when JavaScript is available.

1. Client requests a page including a `<delete-button></delete-button>` tag.
2. Enhance SSR finds the component in `app/components/delete-button.mjs`.
3. Enhance SSR runs the `render` method of the `DeleteButton` class on the server.

> 4. The entire page is returned to the client.

Reviewer: ```suggestion
4. The entire page is returned to the client as HTML and CSS.
```

5. The client encounters a `script` tag and downloads it from the server.

> 6. This `script` tag includes the code for `DeleteButton` so it is evaluated and executed. When executed the `custom-elements.define('delete-button', DeleteButton)` method will be called registering your web component with the browser.

Reviewer: ```suggestion
6. This `script` tag includes the code for `DeleteButton` so it is evaluated and executed. When executed the `custom-elements.define('delete-button', DeleteButton)` method will be called, thus registering your web component with the browser.
```

> This progressively enhances your, presentation only element, into an interactive client side web component.

Reviewer: ```suggestion
This progressively enhances your presentational element into an interactive client side web component.
```

</doc-callout>

### Making Components available in the browser

> Now that we have written a client side component we need to register it with the browser at runtime. To do this we will create a new file `app/browser/index.mjs`. This is the file where we will import all of the components that will be rendered client-side. Using our `MyCard` component as an example our `app/browser/index.mjs` would look like this:

Reviewer: Minor suggestions because I'm pedantic 😅

```suggestion
Now that we have written a client side component, we need to register it with the browser at runtime. To do this, we will create a new file: `app/browser/index.mjs`. This file will be used to import our Components at runtime in order to make them available to the browser. Using our `MyCard` component as an example our `app/browser/index.mjs` would look like this:
```

```javascript
// app/browser/index.mjs
import MyCard from "../components/my-card.mjs"
export { MyCard }
```

> This will create a file called `public/browser/index.mjs` that you include as a `script` tag on any pages you want to use the `MyCard` component. For example:

Reviewer: ```suggestion
This file will be mirrored and made available to your application as `public/browser/index.mjs`. You can then use this file as a `script` tag’s source on any pages in which you want to use the `MyCard` component. For example:
```

Author: Is mirroring the right wording here? It's actually rolling up all the references in that file and bundling them into the public/browser folder.

Reviewer: Hmm yeah mirroring isn't correct then. Maybe… this?

```suggestion
This will create a new client side bundle, available at `public/browser/index.mjs`. You can use this file as a `script` tag’s source on any pages in which you want to use the `MyCard` component. For example:
```

```html
<script type="module" src="/_public/browser/index.mjs"></script>
```

<doc-callout level="info" mark="ℹ️">

> At this point, you may wonder why doesn't Enhance automatically make all components available to the browser at run-time? We've debated this internally and decided that what is best for users is for you to have complete control over which components are available on every page. This way, you can customize each route with a `script` tag that only includes the required components on that page. This reduces the amount of JavaScript needed for each route.

Reviewer: ```suggestion
You may wonder why Enhance doesn’t automatically make all components available to the browser at run-time. This is because we believe application authors should have complete control over which components are available on every page. This way, you can customize each route with a `script` tag that only includes the components required by an individual page. This reduces the amount of JavaScript needed for each route.
```
